### ISSUE TITLE
For HyperbandScheduler, epoch must start with 1, not with 0

### ISSUE THREAD

**Issue by mseeger** [awslabs/autogluon#478]
I noted a mistake in the docstrings, and maybe also in the tabular code.

When time_attr = 'epoch' in HyperbandScheduler, the values coming from the reporter are assumed to start with 1 and run up to max_t. Instead, the examples make them start at 0 and run up to max_t-1.

In other words, HyperbandScheduler treats epoch is "number of epochs done" (which starts at 1). Or think about time or resource spent. The value 0 makes no sense here.

Unfortunately, this really matters. The HB scheduler does certain things when rung levels are reached, and they are 1, 3, 9, ... When train_fn does something like:

for e in range(args.epochs):
    # ...
    reporter(epoch=e, ...)

then the first value returned is not used by the scheduler, but the second is associated with 1. Worse, the final one for max_t is expected by the scheduler, but never comes. This is a problem certainly for the GP methods we are about to contribute.

I could fix this, but I am not very familiar with the code in tabular. I can fix the docstrings.

**Comment by Innixma** [comment]
Would it instead make sense to change the usage of epochs in HyperbandScheduler to use epoch+1, that way it can start at 0? We should try to have consistency between the schedulers and reporter in terms of what these values mean if they are shared between them. I am fine with changes being made in a new PR to address this.

**Comment by mseeger** [comment]
I'd be against changing HyperbandScheduler, and instead insist that epoch must start with 1.

If you think of this more general as a resource attribute, you want to count how much resource (or time) has been spent. So 0 makes no sense.

I think as long as it is clearly said that time_attr (def: epoch) in HyperbandScheduler must be a positive integer, we are fine. No?

**Comment by Innixma** [comment]
I'm fine with this so long as it is documented, and that we confirm it is resolved in all locations (including Tabular) in your PR.

@jwmueller , @zhreshold are we all aligned?

**Comment by jwmueller** [comment]
A simple solution might be: we add a flag-argument to the __init__ of HyperbandScheduler that simply internally increments it's epochs by +1 if the argument is specified?  Each task can then properly set this flag when creating its scheduler.

But AutoGluon's overall epoch-indexing anyway seems poorly unified so it could be worth addressing at this time: 

Image classification uses 1-based epoch indices: https://github.com/awslabs/autogluon/blob/master/autogluon/task/image_classification/pipeline.py#L110

Object detection uses 0-based: https://github.com/awslabs/autogluon/blob/master/autogluon/task/object_detection/object_detection.py#L223

Text classification uses 0-based: https://github.com/awslabs/autogluon/blob/master/autogluon/task/text_classification/pipeline.py#L284

Tabular uses some 0-based in Catboost and NN: https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py#L313

I wouldn't be opposed to switching everything over to 1-based epoch indexing, but some third party packages like Catboost may not support that (anyway "epoch" has a different meaning in catboost so it's ok to have mismatch there IMO).  But I'm not aware of all the downstream things such a switch could possibly break, and whether it may affect the non-Hyperband schedulers.

**Comment by mseeger** [comment]
@jwmueller But would we not control the train_fn wrapper for any 3rd-party code?

I can do what you suggest (in #470), this flag would either leave time_attr values returned by reporter as they are (default) or increment them by 1.

But then, one can ask why this is not just done in the train_fn wrapper.

One other thing (which I think I documented): HyperbandScheduler requires time_attr values to be (positive) int, you cannot use float. This is of course specific to this scheduler (which operates with resource levels, which are int), it could be different for other schedulers.

Let me know what you prefer.

BTW: In #470, I already changed all the docs, and here and there also added an explanation that we want epoch to mean "number of epochs done", and it should start from 1.
